### ISSUE TITLE
"az role assignment list" Not working as expected

### ISSUE THREAD

**Issue by juanjojulian** [Azure/azure-cli#7934]
**Describe the bug**
While trying to list all assignments for a subscription I get the following list:
```
juanjo@Dune roles$ az role assignment list -o table --subscription MySub
Principal                                                                   Role         Scope
--------------------------------------------------------------------------  -----------  ---------------------------------------------------
nice-admins                                                                Owner        /subscriptions/MySubID
pepe@pepo.org                                                               Owner        /subscriptions/MySubID
https://pepo.onmicrosoft.com/568e2906-4e79-6373-876e-7652ec41a893  Contributor  /subscriptions/MySubID
sub-nice-admins                                                           Contributor  /subscriptions/MySubID
                                                                            Contributor  /subscriptions/MysubID
```
Please note that:
1. The App registration we use to deploy with Terraform is listed with the identifier Uri instead of the display name, this is a problem for us because it makes much more difficult to automate things and I believe it shouldn't be the standard behaviour.
2. Last line has the Principal name empty. If I use the Portal IAM I can see the name of the user.

**To Reproduce**
az role assignment list -o table --subscription MySub

**Expected behavior**
List all assignments with correct display names.

**Environment summary**

Installed with brew
Mac OS Mojave
Bash

```
azure-cli (2.0.51)

acr (2.1.9)
acs (2.3.11)
advisor (2.0.0)
ams (0.3.0)
appservice (0.2.7)
backup (1.2.1)
batch (3.4.1)
batchai (0.4.4)
billing (0.2.0)
botservice (0.1.1)
cdn (0.2.0)
cloud (2.1.0)
cognitiveservices (0.2.4)
command-modules-nspkg (2.0.2)
configure (2.0.19)
consumption (0.4.0)
container (0.3.9)
core (2.0.51)
cosmosdb (0.2.5)
dla (0.2.3)
dls (0.1.5)
dms (0.1.1)
eventgrid (0.2.0)
eventhubs (0.3.1)
extension (0.2.3)
feedback (2.1.4)
find (0.2.12)
hdinsight (0.1.0)
interactive (0.4.0)
iot (0.3.4)
iotcentral (0.1.4)
keyvault (2.2.7)
lab (0.1.3)
maps (0.3.2)
monitor (0.2.7)
network (2.2.9)
nspkg (3.0.3)
policyinsights (0.1.0)
profile (2.1.2)
rdbms (0.3.5)
redis (0.3.2)
relay (0.1.2)
reservations (0.4.1)
resource (2.1.6)
role (2.1.10)
search (0.1.1)
servicebus (0.3.2)
servicefabric (0.1.8)
signalr (1.0.0)
sql (2.1.5)
storage (2.2.5)
telemetry (1.0.0)
vm (2.2.8)

Python location '/usr/local/opt/python/bin/python3.7'
Extensions directory '/Users/juanjo/.azure/cliextensions'

Python (Darwin) 3.7.1 (default, Nov  6 2018, 18:46:03) 
[Clang 10.0.0 (clang-1000.11.45.5)]

```

**Additional context**

**Comment by yugangw-msft** [comment]
I suggest you use `--json` which would let you see more things, particularly the `principalId` which would help you find the identifier uri through "az ad sp show --id"
For 2, this is caused by a deleted service principal which still has role assignment leftover.

I am closing this, but do let me know if you are still blocked

**Comment by tjprescott** [comment]
Additionally, you should not be using table output for any kind of programmatic access or scripts. You should use `json` or `csv` format.

**Comment by juanjojulian** [comment]
Thanks @yugangw-msft and @tjprescott . I found other ways to accomplish this but I still think that the output for "az role assignment list -o table --subscription MySub" makes no sense for Apps regs. The Azure CLI is showing the "servicePrincipalNames" which in the case of Apps is just and URI or an ID, its impossible to identify the App with those names, while there is a "appDisplayName" field much better to be listed in a table.

Example:
```
"appDisplayName": "terraform-spoke-subscription-vnet"
"servicePrincipalNames": [
    "https://tenant.onmicrosoft.com/64h78-7g9o-6732-674h-900673ah6",
    "hja7389-kd87-nj78-67h-82910js"
  ],
```

If I use "az role assignment list --subscription MYSub" with the default json output I get much more info but still "appDisplayName" is not showed and the rest of the fields doesn't help to identify the App as all of them are URIs or IDs, we are humans not machines. In my opinion a proper list should have a field easy to understand for humans like "appDisplayName" and a field easy to be treated in a script and for computers like "principalId".

regarding:

> For 2, this is caused by a deleted service principal which still has role assignment leftover.

Not really, I investigated further and this caused by a principal with no name or at least its principalName filed empty:
```
"principalName": "",
```
